overlord/ifacestate/apparmorprompting/common/patterns.go

```diff
@@ -0,0 +1,536 @@
+	worstPriority priorityType = iota
+	priorityGlobDoublestar
+	priorityTerminalDoublestar
+	priorityDoublestar
+	priorityGlob
+	prioritySinglestar
+	prioritySingleChar
+	priorityTerminated
+	priorityLiteral
+)
+
+type nextPatternsContainer struct {
+	currPriority     priorityType
+	nextPatternsList []*strings.Reader
+}
+
+func (np *nextPatternsContainer) addWithPriority(priority priorityType, reader *strings.Reader) {
+	if priority < np.currPriority {
+		return
+	}
+	if priority > np.currPriority {
+		np.nextPatternsList = np.nextPatternsList[:0]
+		np.currPriority = priority
+	}
+	np.nextPatternsList = append(np.nextPatternsList, reader)
+}
+
+func (np *nextPatternsContainer) nextPatterns() []*strings.Reader {
+	return np.nextPatternsList
+}
+
+// GetHighestPrecedencePattern determines which of the given path patterns is
+// the most specific.
+//
+// Assumes that all of the given patterns satisfy ValidatePathPattern(), so this
+// is not verified as part of this function. Additionally, also assumes that the
+// patterns have been previously expanded using ExpandPathPattern(), so there
+// are no groups in any of the patterns.
+//
+// Below are some sample patterns, in order of precedence, though precedence is
+// only guaranteed between two patterns which may match the same path:
+//
+//	# literals
+//	- /foo/bar/baz
+//	- /foo/bar/
+//	# terminated
+//	- /foo/bar
+//	# any single character
+//	- /foo/bar?baz
+//	- /foo/bar?
+//	- /foo/bar?/
+//	# singlestars
+//	- /foo/bar/*/baz
+//	- /foo/bar/*/
+//	- /foo/bar/*/*baz
+//	- /foo/bar/*/*
+//	- /foo/bar/*
+//	- /foo/bar/*/**
+//	# glob
+//	- /foo/bar*baz
+//	- /foo/bar*/baz
+//	- /foo/bar*/baz/**
+//	- /foo/bar*/
+//	- /foo/bar*/*baz
+//	- /foo/bar*/*/baz
+//	- /foo/bar*/*/
+//	- /foo/bar*/*
+//	- /foo/bar*
+//	# doublestars
+//	- /foo/bar/**/baz
+//	- /foo/bar/**/*baz/
+//	- /foo/bar/**/*baz
+//	# terminal doublestar
+//	- /foo/bar/**/        # These are tough... usually, /foo/bar/**/ would have precedence over
+//	- /foo/bar/**/*       # precedence over /foo/bar/**/*baz, but in this case,
+//	- /foo/bar/**         # the trailing *baz adds more specificity.
+//	# glob with immediate doublestar
+//	- /foo/bar*/**/baz
+//	- /foo/bar*/**/
+//	- /foo/bar*/**
+func GetHighestPrecedencePattern(patterns []string) (string, error) {
+	if len(patterns) == 0 {
+		return "", ErrNoPatterns
+	}
+	alreadySeen := make(map[string]bool, len(patterns))
+	patternForReader := make(map[*strings.Reader]string, len(patterns))
+	remainingPatterns := make([]*strings.Reader, 0, len(patterns))
+	for _, pattern := range patterns {
+		if alreadySeen[pattern] {
+			continue
+		}
+		alreadySeen[pattern] = true
+		if strings.HasSuffix(pattern, `\`) && !strings.HasSuffix(pattern, `\\`) {
+			return "", fmt.Errorf(`invalid path pattern: trailing unescaped '\' character: %q`, pattern)
+		}
+		reader := strings.NewReader(pattern)
+		patternForReader[reader] = pattern
+		remainingPatterns = append(remainingPatterns, reader)
+	}
```

Comment:
given that this is meant to be used on patterns that all match the same file, wouldn't it make sense to as first step remove any common prefix from all the patterns?